cmd/kube-proxy/app/server.go

```diff
@@ -822,16 +822,17 @@ func (s *ProxyServer) Run() error {
 
 	// TODO(thockin): make it possible for healthz and metrics to be on the same port.
 
-	var errCh chan error
+	var healthzErrCh, metricsErrCh chan error
 	if s.Config.BindAddressHardFail {
-		errCh = make(chan error)
+		healthzErrCh = make(chan error)
+		metricsErrCh = make(chan error)
 	}
 
 	// Start up a healthz server if requested
-	serveHealthz(s.HealthzServer, errCh)
+	serveHealthz(s.HealthzServer, healthzErrCh)
```

Comment:
So looking at this code some more, it feels like we're exposing weird details of `serveHealthz`'s and `serveMetrics`'s implementation to their caller. I think it would be nicer to just pass the value of `s.Config.BindAddressHardFail` to those functions, and have the functions return an error (which would always be `nil` if `BindAddressHardFail` was `false`), and `Run()` can emit the Event and return immediately if either serve function returns an error. (They would still need to use a channel internally, but they'd just make their own channel.)

Reply:
(And that's a problem with the existing code, not your refactoring. I just hadn't looked carefully at this until you started poking at it :slightly_smiling_face: )